File: sdks/python/src/sandboxer/providers/e2b.py
from __future__ import annotations

import base64
import os
import time
from datetime import datetime
from urllib.parse import quote, urlencode

from ..types import (
    ProviderName,
    SandboxStatus,
    SandboxInfo,
    CommandResult,
    FileInfo,
    ProcessInfo,
    PTYInfo,
    CreateSandboxRequest,
    RunCommandRequest,
    StartCommandRequest,
    CreatePTYRequest,
    ListSandboxesFilter,
)
from ..config import ProviderConfig
from ..provider import Provider, Sandbox
from ..errors import NotFoundError, NotSupportedError, ProviderError, BadConfigError
from .._http_client import HttpClient, HTTPError
from .._connect_proto import unary_post, stream_post, CONNECT_TIMEOUT_MS
from .._util import first_non_empty, shell_quote
from ..registry import register_provider

_DEFAULT_API_BASE = "https://api.e2b.app"
_DEFAULT_ENVD_PORT = 49983
_HEADER_API_KEY = "X-API-Key"
_HEADER_ACCESS_TOKEN = "X-Access-Token"


def _elapsed_ms(start: float) -> int:
    return int((time.monotonic() - start) * 1000)


def _map_err(name: str, err: HTTPError) -> Exception:
    if err.status == 404:
        return NotFoundError(str(err))
    msg = err.body.decode(errors="replace")
    if len(msg) > 256:
        msg = msg[:256] + "..."
    return ProviderError(provider=name, message=msg, status_code=err.status)


def _metadata_match(m: dict[str, str], needle: str) -> bool:
    if not needle:
        return True
    for v in m.values():
        if needle in v:
            return True
    return False


class E2BProvider(Provider):
    def __init__(self, cfg: ProviderConfig) -> None:
        key = first_non_empty(cfg.api_key, os.environ.get("E2B_API_KEY"))
        if not key:
            raise BadConfigError("E2B API key required (api_key or E2B_API_KEY)")
        timeout_s = 30.0
        if cfg.default_timeout_ms:
            timeout_s = cfg.default_timeout_ms / 1000.0
        self._hc = HttpClient(timeout_s=timeout_s)
        self._api_key = key
        self._api_base = (cfg.base_url or _DEFAULT_API_BASE).rstrip("/")
        port_str = os.environ.get("E2B_ENVD_PORT", "")
        self._port = int(port_str) if port_str.isdigit() else _DEFAULT_ENVD_PORT
        self._tpl = os.environ.get("E2B_TEMPLATE_ID", "base")

    def _api_headers(self) -> dict[str, str]:
        return {_HEADER_API_KEY: self._api_key}

    def _envd_base(self, sandbox_id: str) -> str:
        return f"https://{self._port}-{sandbox_id}.e2b.app"

    def _envd_headers(self, token: str | None) -> dict[str, str]:
        h: dict[str, str] = {}
        if token:
            h[_HEADER_ACCESS_TOKEN] = token
        return h

    def _get_sandbox_detail(self, sandbox_id: str) -> dict:
        url = f"{self._api_base}/sandboxes/{quote(sandbox_id, safe='')}"
        try:
            result = self._hc.do("GET", url, self._api_headers())
        except HTTPError as e:
            raise _map_err("e2b", e) from e
        return result  # type: ignore[return-value]

    def _detail_to_info(self, d: dict, fallback_tpl: str, token: str | None = None) -> SandboxInfo:
        tpl = d.get("templateID") or fallback_tpl
        state = d.get("state", "running")
        status = SandboxStatus.PAUSED if state == "paused" else SandboxStatus.RUNNING
        started = datetime.now()
        if d.get("startedAt"):
            try:
                started = datetime.fromisoformat(d["startedAt"].replace("Z", "+00:00"))
            except (ValueError, TypeError):
                pass
        expires: datetime | None = None
        if d.get("endAt"):
            try:
                expires = datetime.fromisoformat(d["endAt"].replace("Z", "+00:00"))
            except (ValueError, TypeError):
                pass
        cpus = d.get("cpuCount") or None
        memory = d.get("memoryMB") or None
        return SandboxInfo(
            id=d.get("sandboxID", ""),
            provider=ProviderName.E2B,
            template=tpl,
            status=status,
            started_at=started,
            expires_at=expires,
            metadata=d.get("metadata") or {},
            cpus=cpus if cpus and cpus > 0 else None,
            memory_mb=memory if memory and memory > 0 else None,
        )

    def list_sandboxes(self, filter: ListSandboxesFilter | None = None) -> list[SandboxInfo]:
        filter = filter or ListSandboxesFilter()
        if filter.provider is not None and filter.provider != ProviderName.E2B:
            return []
        url = f"{self._api_base}/v2/sandboxes"
        if filter.limit and filter.limit > 0:
            url += f"?limit={filter.limit}"
        try:
            listed = self._hc.do("GET", url, self._api_headers())
        except HTTPError as e:
            raise _map_err("e2b", e) from e
        if not listed:
            return []
        out: list[SandboxInfo] = []
        for s in listed:  # type: ignore[union-attr]
            metadata = s.get("metadata") or {}
            if filter.metadata_filter and not _metadata_match(metadata, filter.metadata_filter):
                continue
            state = s.get("state", "running")
            status = SandboxStatus.PAUSED if state == "paused" else SandboxStatus.RUNNING
            started = datetime.now()
            if s.get("startedAt"):
                try:
                    started = datetime.fromisoformat(s["startedAt"].replace("Z", "+00:00"))
                except (ValueError, TypeError):
                    pass
            expires: datetime | None = None
            if s.get("endAt"):
                try:
                    expires = datetime.fromisoformat(s["endAt"].replace("Z", "+00:00"))
                except (ValueError, TypeError):
                    pass
            cpus = s.get("cpuCount") or None
            memory = s.get("memoryMB") or None
            out.append(
                SandboxInfo(
                    id=s.get("sandboxID", ""),
                    provider=ProviderName.E2B,
                    template=s.get("templateID"),
                    status=status,
                    started_at=started,
                    expires_at=expires,
                    metadata=metadata,
                    cpus=cpus if cpus and cpus > 0 else None,
                    memory_mb=memory if memory and memory > 0 else None,
                )
            )
        return out

    def kill_sandbox(self, sandbox_id: str) -> None:
        url = f"{self._api_base}/sandboxes/{quote(sandbox_id, safe='')}"
        try:
            self._hc.do("DELETE", url, self._api_headers())
        except HTTPError as e:
            raise _map_err("e2b", e) from e

    def create_sandbox(self, req: CreateSandboxRequest | None = None) -> tuple[Sandbox, SandboxInfo]:
        req = req or CreateSandboxRequest()
        tpl = self._tpl
        if req.template:
            tpl = req.template
        body: dict = {
            "templateID": tpl,
            "metadata": req.metadata,
            "envVars": req.envs,
        }
        if req.timeout_seconds is not None:
            body["timeout"] = req.timeout_seconds
        if req.cpus is not None:
            body["cpuCount"] = req.cpus
        if req.memory_mb is not None:
            body["memoryMB"] = req.memory_mb
        url = f"{self._api_base}/sandboxes"
        try:
            created = self._hc.do("POST", url, self._api_headers(), body)
        except HTTPError as e:
            raise _map_err("e2b", e) from e
        sandbox_id = created["sandboxID"]  # type: ignore[index]
        token = created.get("envdAccessToken")  # type: ignore[union-attr]
        detail = self._get_sandbox_detail(sandbox_id)
        info = self._detail_to_info(detail, tpl, token)
        sb = E2BSandbox(provider=self, sandbox_id=sandbox_id, token=token)
        return sb, info

    def attach_sandbox(self, sandbox_id: str) -> Sandbox:
        detail = self._get_sandbox_detail(sandbox_id)
        token = detail.get("envdAccessToken")
        return E2BSandbox(provider=self, sandbox_id=sandbox_id, token=token)

    def close(self) -> None:
        self._hc.close()


class E2BSandbox(Sandbox):
    def __init__(self, provider: E2BProvider, sandbox_id: str, token: str | None = None) -> None:
        self._provider = provider
        self._id = sandbox_id
        self._token = token

    @property
    def id(self) -> str:
        return self._id

    def info(self) -> SandboxInfo:
        detail = self._provider._get_sandbox_detail(self._id)
        return self._provider._detail_to_info(detail, self._provider._tpl, self._token)

    def is_running(self) -> bool:
        detail = self._provider._get_sandbox_detail(self._id)
        return detail.get("state") == "running"

    def pause(self) -> None:
        url = f"{self._provider._api_base}/sandboxes/{quote(self._id, safe='')}/pause"
        try:
            self._provider._hc.do("POST", url, self._provider._api_headers(), {})
        except HTTPError as e:
            raise _map_err("e2b", e) from e

    def resume(self) -> None:
        url = f"{self._provider._api_base}/sandboxes/{quote(self._id, safe='')}/resume"
        try:
            self._provider._hc.do("POST", url, self._provider._api_headers(), {})
        except HTTPError as e:
            raise _map_err("e2b", e) from e

    def kill(self) -> None:
        self._provider.kill_sandbox(self._id)

    def port_url(self, port: int) -> str:
        return f"https://{port}-{self._id}.e2b.app"

    def run_command(self, req: RunCommandRequest) -> CommandResult:
        start = time.monotonic()
        url = self._provider._envd_base(self._id) + "/process.Process/Start"
        h = self._provider._envd_headers(self._token)
        if req.timeout_seconds is not None and req.timeout_seconds > 0:
            h[CONNECT_TIMEOUT_MS] = str(req.timeout_seconds * 1000)
        proc: dict = {
            "cmd": "sh",
            "args": ["-c", req.cmd],
            "envs": req.env,
        }
        if req.cwd:
            proc["cwd"] = req.cwd
        body = {"process": proc}
        stdout_parts: list[str] = []
        stderr_parts: list[str] = []
        exit_code = -1
        saw_end = False

        def on_msg(msg: dict) -> None:
            nonlocal exit_code, saw_end
            event = msg.get("event")
            if not isinstance(event, dict):
                return
            data = event.get("data")
            if isinstance(data, dict):
                _append_b64(stdout_parts, data.get("stdout"))
                _append_b64(stderr_parts, data.get("stderr"))
            end = event.get("end")
            if isinstance(end, dict):
                ec = end.get("exitCode", end.get("exit_code", 0))
                exit_code = ec if isinstance(ec, int) else 0
                saw_end = True

        stream_post(self._provider._hc, url, h, body, on_msg)
        if not saw_end:
            exit_code = -1
        return CommandResult(
            stdout="".join(stdout_parts),
            stderr="".join(stderr_parts),
            exit_code=exit_code,
            duration_ms=_elapsed_ms(start),
        )

    def start_command(self, req: StartCommandRequest) -> tuple[int, str]:
        raise NotSupportedError("E2B does not support start_command")

    def wait_for_handle(self, handle_id: str) -> CommandResult:
        raise NotSupportedError("E2B does not support wait_for_handle")

    def kill_process(self, pid: int) -> None:
        raise NotSupportedError("E2B does not support kill_process")

    def list_processes(self) -> list[ProcessInfo]:
        raise NotSupportedError("E2B does not support list_processes")

    def read_file(self, path: str) -> bytes:
        url = self._provider._envd_base(self._id) + "/files?" + urlencode({"path": path})
        h = self._provider._envd_headers(self._token)
        try:
            _, data = self._provider._hc.do_raw("GET", url, h)
        except HTTPError as e:
            if e.status == 404:
                raise NotFoundError(f"file not found: {path}") from e
            raise _map_err("e2b", e) from e
        return data

    def write_file(self, path: str, content: bytes, mode: int | None = None, user: str | None = None) -> None:
        url = self._provider._envd_base(self._id) + "/files"
        h = self._provider._envd_headers(self._token)
        # multipart upload
        import io

        boundary = "----SandboxerBoundary"
        body = io.BytesIO()
        body.write(f"--{boundary}\r\n".encode())
        body.write(b'Content-Disposition: form-data; name="path"\r\n\r\n')
        body.write(path.encode())
        body.write(b"\r\n")
        body.write(f"--{boundary}\r\n".encode())
        body.write(b'Content-Disposition: form-data; name="file"; filename="blob"\r\n')
        body.write(b"Content-Type: application/octet-stream\r\n\r\n")
        body.write(content)
        body.write(b"\r\n")
        body.write(f"--{boundary}--\r\n".encode())
        h["Content-Type"] = f"multipart/form-data; boundary={boundary}"
        try:
            self._provider._hc.do_raw("POST", url, h, content=body.getvalue())
        except HTTPError as e:
            raise _map_err("e2b", e) from e

    def list_directory(self, path: str) -> list[FileInfo]:
        url = self._provider._envd_base(self._id) + "/filesystem.Filesystem/ListDir"
        h = self._provider._envd_headers(self._token)
        result = unary_post(self._provider._hc, url, h, {"path": path, "depth": 1})
        if not result:
            return []
        entries = result.get("entries", []) if isinstance(result, dict) else []
        out: list[FileInfo] = []
        for e in entries:
            is_dir = "DIRECTORY" in (e.get("type", "") or "")
            size = 0
            raw_size = e.get("size")
            if isinstance(raw_size, (int, float)):
                size = int(raw_size)
            elif isinstance(raw_size, str) and raw_size.isdigit():
                size = int(raw_size)
            file_mode = e.get("mode") or None
            out.append(
                FileInfo(
                    name=e.get("name", ""),
                    path=e.get("path", ""),
                    is_dir=is_dir,
                    size=size,
                    mode=file_mode if file_mode and file_mode != 0 else None,
                )
            )
        return out

    def make_dir(self, path: str) -> None:
        url = self._provider._envd_base(self._id) + "/filesystem.Filesystem/MakeDir"
        h = self._provider._envd_headers(self._token)
        unary_post(self._provider._hc, url, h, {"path": path})

    def remove(self, path: str) -> None:
        self.run_command(RunCommandRequest(cmd="rm -rf " + shell_quote(path)))

    def exists(self, path: str) -> bool:
        url = self._provider._envd_base(self._id) + "/files?" + urlencode({"path": path})
        h = self._provider._envd_headers(self._token)
        try:
            self._provider._hc.do_raw("GET", url, h)
        except HTTPError as e:
            if e.status == 404:
                return False
            raise _map_err("e2b", e) from e
        return True

    def create_pty(self, req: CreatePTYRequest) -> PTYInfo:
        raise NotSupportedError("E2B does not support create_pty")

    def resize_pty(self, pid: int, rows: int, cols: int) -> None:
        raise NotSupportedError("E2B does not support resize_pty")

    def kill_pty(self, pid: int) -> None:
        raise NotSupportedError("E2B does not support kill_pty")

    def list_pty(self) -> list[PTYInfo]:
        raise NotSupportedError("E2B does not support list_pty")


def _append_b64(parts: list[str], val: str | None) -> None:
    if not val:
        return
    try:
        decoded = base64.b64decode(val).decode(errors="replace")
        parts.append(decoded)
    except Exception:
        parts.append(val)


def _make_e2b_provider(cfg: ProviderConfig) -> Provider:
    return E2BProvider(cfg)


register_provider(ProviderName.E2B, _make_e2b_provider)

File: sdks/typescript/src/providers/e2b.ts
import type { ProviderConfig } from "../config.js";
import type { Provider, Sandbox } from "../provider.js";
import type {
  SandboxInfo,
  CommandResult,
  FileInfo,
  ProcessInfo,
  PTYInfo,
  CreateSandboxRequest,
  RunCommandRequest,
  StartCommandRequest,
  CreatePTYRequest,
  ListSandboxesFilter,
} from "../types.js";
import { registerProvider } from "../registry.js";
import { HttpClient, HTTPError } from "../http-client.js";
import { unaryPost, streamPost } from "../connect-proto.js";
import { NotFoundError, NotSupportedError, ProviderError } from "../errors.js";
import { firstNonEmpty, shellQuote } from "../util.js";

const DEFAULT_API_BASE = "https://api.e2b.app";
const DEFAULT_ENVD_PORT = 49983;
const HEADER_API_KEY = "X-API-Key";
const HEADER_ACCESS_TOKEN = "X-Access-Token";

class E2BProvider implements Provider {
  private hc: HttpClient;
  private apiKey: string;
  private apiBase: string;
  private port: number;
  private tpl: string;

  constructor(
    hc: HttpClient,
    apiKey: string,
    apiBase: string,
    port: number,
    tpl: string,
  ) {
    this.hc = hc;
    this.apiKey = apiKey;
    this.apiBase = apiBase;
    this.port = port;
    this.tpl = tpl;
  }

  private apiHeaders(): Record<string, string> {
    return { [HEADER_API_KEY]: this.apiKey };
  }

  envdBase(sandboxId: string): string {
    return `https://${this.port}-${sandboxId}.e2b.app`;
  }

  envdHeaders(token?: string): Record<string, string> {
    const h: Record<string, string> = {};
    if (token) h[HEADER_ACCESS_TOKEN] = token;
    return h;
  }

  async listSandboxes(filter?: ListSandboxesFilter): Promise<SandboxInfo[]> {
    if (filter?.provider && filter.provider !== "e2b") return [];
    let u = `${this.apiBase}/v2/sandboxes`;
    if (filter?.limit && filter.limit > 0) u += `?limit=${filter.limit}`;

    const listed = await this.hc.do<
      Array<{
        templateID: string;
        sandboxID: string;
        startedAt: string;
        endAt: string;
        state: string;
        metadata: Record<string, string>;
        cpuCount: number;
        memoryMB: number;
      }>
    >("GET", u, this.apiHeaders());

    const out: SandboxInfo[] = [];
    for (const s of listed ?? []) {
      if (
        filter?.metadataFilter &&
        !metadataMatch(s.metadata, filter.metadataFilter)
      )
        continue;
      const info: SandboxInfo = {
        id: s.sandboxID,
        provider: "e2b",
        template: s.templateID,
        status: s.state === "paused" ? "paused" : "running",
        startedAt: s.startedAt || new Date().toISOString(),
        metadata: s.metadata,
      };
      if (s.endAt) info.expiresAt = s.endAt;
      if (s.cpuCount > 0) info.cpus = s.cpuCount;
      if (s.memoryMB > 0) info.memoryMb = s.memoryMB;
      out.push(info);
    }
    return out;
  }

  async killSandbox(sandboxId: string): Promise<void> {
    const u = `${this.apiBase}/sandboxes/${encodeURIComponent(sandboxId)}`;
    try {
      await this.hc.do("DELETE", u, this.apiHeaders());
    } catch (e) {
      throw mapE2BErr(e);
    }
  }

  async createSandbox(
    req?: CreateSandboxRequest,
  ): Promise<[Sandbox, SandboxInfo]> {
    const tpl = req?.template || this.tpl;
    const body: Record<string, unknown> = {
      templateID: tpl,
      metadata: req?.metadata,
      envVars: req?.envs,
    };
    if (req?.timeoutSeconds) body.timeout = req.timeoutSeconds;
    if (req?.cpus) body.cpuCount = req.cpus;
    if (req?.memoryMb) body.memoryMB = req.memoryMb;

    let created: {
      sandboxID: string;
      templateID: string;
      envdAccessToken?: string;
      trafficAccessToken?: string;
    };
    try {
      created = await this.hc.do(
        "POST",
        `${this.apiBase}/sandboxes`,
        this.apiHeaders(),
        body,
      );
    } catch (e) {
      throw mapE2BErr(e);
    }

    const detail = await this.getSandboxDetail(created.sandboxID);
    const info = detailToInfo(detail, tpl);
    const sb = new E2BSandbox(this, created.sandboxID, created.envdAccessToken);
    return [sb, info];
  }

  async attachSandbox(sandboxId: string): Promise<Sandbox> {
    const detail = await this.getSandboxDetail(sandboxId);
    return new E2BSandbox(this, sandboxId, detail.envdAccessToken);
  }

  async close(): Promise<void> {}

  async getSandboxDetail(id: string): Promise<E2BSandboxDetail> {
    const u = `${this.apiBase}/sandboxes/${encodeURIComponent(id)}`;
    try {
      return await this.hc.do<E2BSandboxDetail>("GET", u, this.apiHeaders());
    } catch (e) {
      throw mapE2BErr(e);
    }
  }

  getHc(): HttpClient {
    return this.hc;
  }
  getTpl(): string {
    return this.tpl;
  }
}

interface E2BSandboxDetail {
  sandboxID: string;
  templateID: string;
  startedAt: string;
  endAt: string;
  state: string;
  metadata: Record<string, string>;
  cpuCount: number;
  memoryMB: number;
  envdAccessToken?: string;
}

function detailToInfo(d: E2BSandboxDetail, fallbackTpl: string): SandboxInfo {
  const tpl = d.templateID || fallbackTpl;
  const info: SandboxInfo = {
    id: d.sandboxID,
    provider: "e2b",
    template: tpl,
    status: d.state === "paused" ? "paused" : "running",
    startedAt: d.startedAt || new Date().toISOString(),
    metadata: d.metadata,
  };
  if (d.endAt) info.expiresAt = d.endAt;
  if (d.cpuCount > 0) info.cpus = d.cpuCount;
  if (d.memoryMB > 0) info.memoryMb = d.memoryMB;
  return info;
}

class E2BSandbox implements Sandbox {
  readonly id: string;
  private provider: E2BProvider;
  private token?: string;

  constructor(provider: E2BProvider, id: string, token?: string) {
    this.provider = provider;
    this.id = id;
    this.token = token;
  }

  async info(): Promise<SandboxInfo> {
    const d = await this.provider.getSandboxDetail(this.id);
    return detailToInfo(d, this.provider.getTpl());
  }

  async isRunning(): Promise<boolean> {
    const d = await this.provider.getSandboxDetail(this.id);
    return d.state === "running";
  }

  async pause(): Promise<void> {
    const u = `${(this.provider as E2BProvider)["apiBase"]}/sandboxes/${encodeURIComponent(this.id)}/pause`;
    try {
      await this.provider
        .getHc()
        .do(
          "POST",
          u,
          { [HEADER_API_KEY]: (this.provider as E2BProvider)["apiKey"] },
          {},
        );
    } catch (e) {
      throw mapE2BErr(e);
    }
  }

  async resume(): Promise<void> {
    const u = `${(this.provider as E2BProvider)["apiBase"]}/sandboxes/${encodeURIComponent(this.id)}/resume`;
    try {
      await this.provider
        .getHc()
        .do(
          "POST",
          u,
          { [HEADER_API_KEY]: (this.provider as E2BProvider)["apiKey"] },
          {},
        );
    } catch (e) {
      throw mapE2BErr(e);
    }
  }

  async kill(): Promise<void> {
    return this.provider.killSandbox(this.id);
  }

  async portUrl(port: number): Promise<string> {
    return `https://${port}-${this.id}.e2b.app`;
  }

  async runCommand(req: RunCommandRequest): Promise<CommandResult> {
    const start = Date.now();
    const u = this.provider.envdBase(this.id) + "/process.Process/Start";
    const h = this.provider.envdHeaders(this.token);
    if (req.timeoutSeconds && req.timeoutSeconds > 0) {
      h["Connect-Timeout-Ms"] = String(req.timeoutSeconds * 1000);
    }
    const proc: Record<string, unknown> = {
      cmd: "sh",
      args: ["-c", req.cmd],
      envs: req.env,
    };
    if (req.cwd) proc.cwd = req.cwd;
    const body = { process: proc };

    let stdout = "";
    let stderr = "";
    let exit = 0;
    let sawEnd = false;

    await streamPost(this.provider.getHc(), u, h, body, (msg: unknown) => {
      const m = msg as {
        event?: {
          data?: { stdout?: string; stderr?: string };
          end?: { exit_code?: number; exitCode?: number };
        };
      };
      if (m.event?.data) {
        if (m.event.data.stdout) {
          try {
            stdout += atob(m.event.data.stdout);
          } catch {
            stdout += m.event.data.stdout;
          }
        }
        if (m.event.data.stderr) {
          try {
            stderr += atob(m.event.data.stderr);
          } catch {
            stderr += m.event.data.stderr;
          }
        }
      }
      if (m.event?.end) {
        const ec = m.event.end.exitCode ?? m.event.end.exit_code ?? 0;
        exit = ec;
        sawEnd = true;
      }
    });

    if (!sawEnd) exit = -1;

    return {
      stdout,
      stderr,
      exitCode: exit,
      durationMs: Date.now() - start,
    };
  }

  async startCommand(
    _req: StartCommandRequest,
  ): Promise<{ pid: number; handleId: string }> {
    throw new NotSupportedError();
  }

  async waitForHandle(_handleId: string): Promise<CommandResult> {
    throw new NotSupportedError();
  }

  async killProcess(_pid: number): Promise<void> {
    throw new NotSupportedError();
  }

  async listProcesses(): Promise<ProcessInfo[]> {
    throw new NotSupportedError();
  }

  async readFile(path: string): Promise<Uint8Array> {
    const u =
      this.provider.envdBase(this.id) +
      "/files?path=" +
      encodeURIComponent(path);
    const h = this.provider.envdHeaders(this.token);
    try {
      const resp = await this.provider.getHc().doRaw("GET", u, h);
      return resp.body;
    } catch (e) {
      if (e instanceof HTTPError && e.status === 404) throw new NotFoundError();
      throw mapE2BErr(e);
    }
  }

  async writeFile(
    path: string,
    content: Uint8Array,
    _mode?: number,
    _user?: string,
  ): Promise<void> {
    const u = this.provider.envdBase(this.id) + "/files";
    const h = this.provider.envdHeaders(this.token);
    const boundary = "----SandboxerBoundary" + Date.now().toString(36);
    const encoder = new TextEncoder();
    const parts: Uint8Array[] = [];

    // path field
    parts.push(
      encoder.encode(
        `--${boundary}\r\nContent-Disposition: form-data; name="path"\r\n\r\n${path}\r\n`,
      ),
    );
    // file field
    parts.push(
      encoder.encode(
        `--${boundary}\r\nContent-Disposition: form-data; name="file"; filename="blob"\r\nContent-Type: application/octet-stream\r\n\r\n`,
      ),
    );
    parts.push(content);
    parts.push(encoder.encode(`\r\n--${boundary}--\r\n`));

    let totalLen = 0;
    for (const p of parts) totalLen += p.length;
    const body = new Uint8Array(totalLen);
    let offset = 0;
    for (const p of parts) {
      body.set(p, offset);
      offset += p.length;
    }

    try {
      await this.provider
        .getHc()
        .doRaw("POST", u, h, body, `multipart/form-data; boundary=${boundary}`);
    } catch (e) {
      throw mapE2BErr(e);
    }
  }

  async listDirectory(path: string): Promise<FileInfo[]> {
    const u =
      this.provider.envdBase(this.id) + "/filesystem.Filesystem/ListDir";
    const h = this.provider.envdHeaders(this.token);
    const out = await unaryPost<{
      entries?: Array<{
        name: string;
        path: string;
        type: string;
        size: number | string;
        mode?: number;
      }>;
    }>(this.provider.getHc(), u, h, { path, depth: 1 });
    const res: FileInfo[] = [];
    for (const e of out?.entries ?? []) {
      let size = 0;
      if (typeof e.size === "number") size = e.size;
      else if (typeof e.size === "string") size = parseInt(e.size, 10) || 0;
      res.push({
        name: e.name,
        path: e.path,
        isDir: (e.type || "").includes("DIRECTORY"),
        size,
        mode: e.mode || undefined,
      });
    }
    return res;
  }

  async makeDir(path: string): Promise<void> {
    const u =
      this.provider.envdBase(this.id) + "/filesystem.Filesystem/MakeDir";
    const h = this.provider.envdHeaders(this.token);
    await unaryPost(this.provider.getHc(), u, h, { path });
  }

  async remove(path: string): Promise<void> {
    await this.runCommand({ cmd: "rm -rf " + shellQuote(path) });
  }

  async exists(path: string): Promise<boolean> {
    const u =
      this.provider.envdBase(this.id) +
      "/files?path=" +
      encodeURIComponent(path);
    const h = this.provider.envdHeaders(this.token);
    try {
      await this.provider.getHc().doRaw("GET", u, h);
      return true;
    } catch (e) {
      if (e instanceof HTTPError && e.status === 404) return false;
      throw mapE2BErr(e);
    }
  }

  async createPty(_req: CreatePTYRequest): Promise<PTYInfo> {
    throw new NotSupportedError();
  }
  async resizePty(_pid: number, _rows: number, _cols: number): Promise<void> {
    throw new NotSupportedError();
  }
  async killPty(_pid: number): Promise<void> {
    throw new NotSupportedError();
  }
  async listPty(): Promise<PTYInfo[]> {
    throw new NotSupportedError();
  }
}

function metadataMatch(
  m: Record<string, string> | undefined,
  needle: string,
): boolean {
  if (!needle || !m) return !needle;
  for (const v of Object.values(m)) {
    if (v.includes(needle)) return true;
  }
  return false;
}

function mapE2BErr(e: unknown): Error {
  if (e instanceof HTTPError) {
    if (e.status === 404) return new NotFoundError();
    const msg = new TextDecoder().decode(e.body).slice(0, 256);
    return new ProviderError("e2b", msg, e.status);
  }
  return e instanceof Error ? e : new Error(String(e));
}

registerProvider("e2b", async (config: ProviderConfig) => {
  const key = firstNonEmpty(config.apiKey, process.env.E2B_API_KEY);
  if (!key)
    throw new ProviderError(
      "e2b",
      "E2B API key required (config.apiKey or E2B_API_KEY)",
    );
  const base = (config.baseUrl || DEFAULT_API_BASE).replace(/\/+$/, "");
  let port = DEFAULT_ENVD_PORT;
  const envPort = process.env.E2B_ENVD_PORT;
  if (envPort) {
    const n = parseInt(envPort, 10);
    if (!isNaN(n)) port = n;
  }
  const tpl = process.env.E2B_TEMPLATE_ID || "base";
  const hc = new HttpClient(config.defaultTimeoutMs);
  return new E2BProvider(hc, key, base, port, tpl);
});
